Next message from tree (not switch-cases)

Y-Pazniak opened the pull request:
New option - loading from tree - has been added. Without placeholders for now

Y-Pazniak committed:
first try to get messages from tree

Y-Pazniak committed:
price from placeholders has been added

Y-Pazniak committed:
m1 (passenger cars) have been filled

MaximPilipeiko commented:
Could you please change base branch to `self-propelled-vehicles`. For ability to see real changes and avoid review the same change twice.

MaximPilipeiko commented on src/main/resources/tree.json:
```diff
@@ -148,17 +172,20 @@
                         {
                           "key": "LESS_3_YEARS_AGE",
                           "nextStep": "FAREWELL",
                           "price": "<b>32 727,55 </b>",
```
```suggestion
                          "price": 32727.55
```

Y-Pazniak replied:
Done.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/tree/Node.java:
```diff
@@ -15,15 +15,18 @@ public class Node {
     private final Command key;
     private final Step nextStep;
+    private final String nextMessage;
     private final String price;
```
```suggestion
    private final Double price;
```

Y-Pazniak replied:
Done.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/tree/Node.java:
```diff
@@ -15,15 +15,18 @@ public class Node {
     private final Command key;
     private final Step nextStep;
+    private final String nextMessage;
     private final String price;
```
```suggestion
    private final Price price;
```

Y-Pazniak replied:
Done.

MaximPilipeiko commented on src/main/resources/tree.json:
```diff
@@ -2,32 +2,38 @@
                 {
                   "key": "LESS_3_YEARS_AGE",
                   "nextStep": "FAREWELL",
                   "price": "<b>544,5</b>",
```
```suggestion
                  "price": "544,5 BYN",
```

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/service/MessagesCreator.java:
```diff
@@ -35,43 +33,17 @@ public static MessagesCreator getInstance() {
+        String nextMessage;
+        try {
+            nextMessage = bundle.getString(userProgress.getNextMessage());
+        } catch (NullPointerException e) {
```
What would happen if another exception were thrown? Maybe better to catch more general exception?
```suggestion
        } catch (Exception e) {
```

Y-Pazniak replied:
Done, but I would like to ask you about  explanations. 
We are requesting a string from a bundle and we can receive this string (if it exists in bundle) or we can not receive a string (if there is not such string in bundle). In my opinion another exceptions here are redundant - the only possible exception here is the NPE.

MaximPilipeiko replied:
If you are sure that only NPE is possible, you can leave NullPointerException of course.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/service/UserProgressManager.java:
```diff
@@ -28,6 +28,18 @@ public void createNewUserProgress(final String chatID) throws UtilsborException
+        final UserProgress userProgress;
+        userProgress = userProgressStorage.get(chatID);
+
+        if (!CommandTree.validateCommand(requestingCommand, userProgress)) {
```
It's better if boolean is return type.
```suggestion
        if (!CommandTree.isValidCommand(requestingCommand, userProgress)) {
```

Y-Pazniak replied:
Done.

Y-Pazniak committed:
Fixed errors with incorrect price message sending.
Html markup removed from the tree.
The tree was prepared hor changing from String to Double

Y-Pazniak committed:
String price replaced to Double price

Y-Pazniak committed:
double price replacd to Price price (Price.class)

Y-Pazniak committed:
wrong method command validation naming changed (validateCommand -> isValidCommand).

Y-Pazniak committed:
Exception is expanded for a bundle request

Y-Pazniak committed:
Map fieldToCommand was replaced by code from tree and enums.

Y-Pazniak committed:
Map fieldToCommand was replaced by code from tree and enums.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/Price.java:
```diff
@@ -0,0 +1,26 @@
+import by.custom.utilcalculator.domain.constants.CurrencyType;
+
+public class Price {
+    private final double amount;
```
Think about using `BigDecimal` instead double. It's not a fact that it's really needed.
https://medium.com/@cancerian0684/which-data-type-would-you-choose-for-storing-currency-values-like-trading-price-dd7489e7a439

Y-Pazniak replied:
Thanks for the article. I understood, that for finance we should use BigDecimal for accuracy of calculations (and changed double to BigDecimal :)

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/constants/CurrencyType.java:
```diff
@@ -0,0 +1,15 @@
+package by.custom.utilcalculator.domain.constants;
+
+public enum CurrencyType {
```
Where it's possibly better to avoid words that don't give additional sense.
```suggestion
public enum Currency {
```

Y-Pazniak replied:
Done.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/tree/CommandTree.java:
```diff
@@ -9,14 +10,13 @@
 public class CommandTree {
-    private final Map<StepsIndicator, Command> fieldsToCommands;
-    private final Map<Command, List<StepsIndicator>> commandsToFields;
+    //private final Map<StepsIndicator, Command> fieldsToCommands;
```
Add comments why it's commented

Y-Pazniak replied:
In this case - we refused using maps, now we use nodes to build assotiations between COMMAND and STEP. Don't need maps anymore. Probably, I saved this maps for safety - if something go wrong in future with big branches, I can return maps easily. But it seems everything works fine - I guess, we can delete this maps.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/tree/ModifierTree.java:
```diff
@@ -46,7 +47,7 @@ public static String getPrice(final UserProgress userProgress) throws UtilsborCo
             }
         }
-        return "error during calculation";
+        return null;
```
good

Y-Pazniak committed:
double replaced by BigDecimal
useless map initialization removed
